### ISSUE TITLE
Split up inactivity_timeout into two timeouts for more granularity

### ISSUE THREAD

**Issue by kraih** [mojolicious/mojo#1492]
Everyone's web apps are different. Some of us need our servers to shut down more quickly, others need extra safety from connection close race conditions. Right now `Mojo::Server::Daemon` only has one `inactivity_timeout` setting that applies to connections with active transactions, as well as keep-alive connections without a transaction.

The default `inactivity_timeout` is `15` seconds, which works reasonably well. But many users end up setting the value much higher because their HTTP requests and WebSocket connections can take much longer. Probably using the global setting, affecting all connections. But even connection specific settings (like for WebSockets) are rarely reset, making server restarts very slow.

Now if we were to split up the global timeout settings into `inactivity_timeout` (for connections with active transactions) and `keep_alive_timeout` (for connections without active transactions), and then have `Mojo::Server::Daemon` switch automatically between the two, it could be a lot more user friendly.

I believe giving everyone two settings to tweak, might in many cases even eliminate the need for `$c->inactivity_timeout(300)`.

**Comment by kraih** [comment]
It might be necessary to optimise the lower level bits of `Mojo::IOLoop::Stream::timeout` a bit for more frequent timeout changes (to avoid performance regressions).

**Comment by kraih** [comment]
Testing this feature should be very simple though, just check the stream `inactivity_timeout` at various points during the request life cycle.

**Comment by kraih** [comment]
I've made a branch with a possible `Mojo::IOLoop::Stream::timeout` optimisation. https://github.com/mojolicious/mojo/commit/ffedf7e3ac28

With patch applied:
```
$ perl -Ilib -Mojo -E 'a(sub ($c) { $c->inactivity_timeout(30); $c->render(text => "Hello Mojo!") })->start' daemon -m production
Server available at http://127.0.0.1:3000

$ wrk -c 20 -d 10 http://127.0.0.1:3000
Running 10s test @ http://127.0.0.1:3000
  2 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     9.66ms    0.96ms  23.77ms   93.89%
    Req/Sec     1.04k    43.88     1.11k    64.50%
  20761 requests in 10.05s, 3.05MB read
Requests/sec:   2066.38
Transfer/sec:    311.13KB
```

Without patch applied:
```
$ perl -Mojo -E 'a(sub ($c) { $c->inactivity_timeout(30); $c->render(text => "Hello Mojo!") })->start' daemon -m production
Server available at http://127.0.0.1:3000

$ wrk -c 20 -d 10 http://127.0.0.1:3000
Running 10s test @ http://127.0.0.1:3000
  2 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    10.16ms    0.91ms  20.89ms   93.22%
    Req/Sec     0.99k    41.46     1.10k    73.00%
  19736 requests in 10.04s, 2.90MB read
Requests/sec:   1965.20
Transfer/sec:    295.88KB
```
Benchmark results seem fairly consistent.

**Comment by kraih** [comment]
With the patch from #1493 i think it's as simple as:
```diff
diff --git a/lib/Mojo/Server/Daemon.pm b/lib/Mojo/Server/Daemon.pm
index b65b061c2..3e2abb9c1 100644
--- a/lib/Mojo/Server/Daemon.pm
+++ b/lib/Mojo/Server/Daemon.pm
@@ -13,8 +13,9 @@ use constant DEBUG => $ENV{MOJO_SERVER_DEBUG} || 0;
 
 has acceptors => sub { [] };
 has [qw(backlog max_clients silent)];
-has inactivity_timeout => sub { $ENV{MOJO_INACTIVITY_TIMEOUT} // 15 };
+has inactivity_timeout => sub { $ENV{MOJO_INACTIVITY_TIMEOUT} // 30 };
 has ioloop             => sub { Mojo::IOLoop->singleton };
+has keep_alive_timeout => sub { $ENV{MOJO_KEEP_ALIVE_TIMEOUT} // 15 };
 has listen       => sub { [split ',', $ENV{MOJO_LISTEN} || 'http://*:3000'] };
 has max_requests => 100;
 
@@ -78,7 +79,8 @@ sub _build_tx {
 
   my $tx = $self->build_tx->connection($id);
   $tx->res->headers->server('Mojolicious (Perl)');
-  my $handle = $self->ioloop->stream($id)->handle;
+  my $handle
+    = $self->ioloop->stream($id)->timeout($self->inactivity_timeout)->handle;
   unless ($handle->isa('IO::Socket::UNIX')) {
     $tx->local_address($handle->sockhost)->local_port($handle->sockport);
     $tx->remote_address($handle->peerhost)->remote_port($handle->peerport);
@@ -155,7 +157,8 @@ sub _finish {
   return $self->_remove($id) if $tx->error || !$tx->keep_alive;
 
   # Build new transaction for leftovers
-  return unless length(my $leftovers = $tx->req->content->leftovers);
+  return $self->ioloop->stream($id)->timeout($self->keep_alive_timeout)
+    unless length(my $leftovers = $tx->req->content->leftovers);
   $tx = $c->{tx} = $self->_build_tx($id, $c);
   $tx->server_read($leftovers);
 }
@@ -192,7 +195,7 @@ sub _listen {
 
       $self->{connections}{$id} = {tls => $tls};
       warn "-- Accept $id (@{[$stream->handle->peerhost]})\n" if DEBUG;
-      $stream->timeout($self->inactivity_timeout);
+      $stream->timeout($self->keep_alive_timeout);
 
       $stream->on(close => sub { $self && $self->_close($id) });
       $stream->on(error =>
```

**Comment by kraih** [comment]
Resolved with #1493.
